README.md
atd-tinymce
===========

### After the Deadline for TinyMCE

[After the Deadline](http://www.afterthedeadline.com) is an [open source](http://open.afterthedeadline.com) software service that checks [spelling, style, and grammar.](http://www.afterthedeadline.com/features.slp).
This project contains an AtD TinyMCE Plugin and an example for using After the Deadline with [TinyMCE](http://tinymce.moxiecode.com).

### Quick Start

1.  Extract the contents of this zip file into the tiny_mce/plugins directory.
2.  You'll need to make up an API key. Ideally your application name followed by a string that is unique to this user is best. There is a performance benefit to using the same key for subsequent requests.
3.  You'll need to set three variables in your TinyMCE init:
4.  You'll need a proxy script on your server. Use the included proxy.php if you can. Otherwise you'll need to write your own.
5.  As a final note, make sure your webpage is encoded in UTF-8 format. AJAX requests use the encoding of the parent website and AtD expects UTF-8. This is important as AtD has better support for accented characters and we're working to support more languages.

### Examples

If you've unzipped this archive on a webserver with PHP installed, you should be able to access [demo.html](demo.html) directly and see everything work.

*   [demo.html](demo.html)

### Commercial use and running your own server

This software requires a running instance of an AtD server.  Automattic operates an instance that you can use for personal projects as long as you don't send too much traffic.  This library are configured to use this server by default.

For high volume and commercial uses of AtD, you must run your own server.  The code is available on Github: [After the Deadline Server](https://github.com/automattic/atd-server).  See the [After the Deadline Developer's page](http://open.afterthedeadline.com/) for more information, and check out the [AtD Developers Google Group](http://groups.google.com/group/atd-developers) for discussion and community support.  

When you run your own server, replace `service.afterthedeadline.com` with your server's hostname.

### Ignore Phrases Capability (Optional)

After the Deadline lets you specify a list of phrases it should not highlight. This is the AtD equivalent of add to dictionary functionality in other programs.
You can hardcode these values in the `atd_ignore_strings` value in TinyMCE init. 
The format is a comma separated list of phrases.  (See example below for more).

Users can choose to ignore phrases by selecting _Ignore always_ when clicking 
a highlighted phrase.  You have to enable this feature for the _Ignore always_
menu to show up.  Do so by setting the `atd_ignore_enable` to the string `"true"`
in TinyMCE init.

If you see _Ignore all_ when clicking an error then this feature is disabled.

#### Manage Ignored Phrases with Cookies

There is a caveat to this client side ignore capability.  You get to create the
user interface for unignoring rules.  The phrases are stored in a cookie named
`atd_ignore`.  This cookie was created with:

`tinymce.util.Cookie.setHash("atd_ignore", ...);`

The format of the cookie is _some+phrase=1&amp;someWord=1&amp;..._.  See the [TinyMCE API](http://tinymce.moxiecode.com/js/tinymce/docs/api/index.html#)
for [tinymce.util.Cookie](http://tinymce.moxiecode.com/js/tinymce/docs/api/index.html#class_tinymce.util.Cookie.html).

[support/atdphrases.js](support/atdphrases.js) and [support/unignore.html](support/unignore.html) are included in this archive to assist you.

#### Manage Ignored Phrases Yourself

The best option is to manage the ignored phrases preference on your server. We do this in our [WordPress 
plugin](http://wordpress.org/extend/plugins/after-the-deadline/). When we generate the TinyMCE init we pull the user's ignore preferences from our database and populate the `atd_ignore_strings` value with them.

The AtD TinyMCE plugin can be configured to make an AJAX call to your server when a user selects _Ignore Always_. To enable this set `atd_ignore_rpc_url` in the 
TinyMCE init. When set AtD will do a GET call to this URL with _the ignored phrase&key=api key_ appended.

### Error Categories

This TinyMCE plugin only shows grammar, spelling, and misused word errors by default.  All other categories of errors are filtered and you must explicitly
enable them in the TinyMCE init parameters by setting the `atd_show_types` variable.

`atd_show_types: "Bias Language,Cliches,Complex Expression,Diacritical Marks,Double Negatives,Hidden Verbs,Jargon Language,Passive voice,Phrases to Avoid,Redundant Expression",`

You may omit any of these categories.  Note that categories are separated by commas with no whitespace.  The category names are case sensitive and yes, voice is lowercase in Passive voice.

### Localization

To localize the strings in this extension, create an object with the localized strings. Here is an example:

<pre>var my_plugin_strings = {
   menu_title_spelling: "Spelling",
   menu_title_repeated_word: "Repeated	Word",
   menu_option_explain: "Explain...",
   menu_option_ignore_once: "Ignore suggestion",
   menu_option_ignore_always: "Ignore always",
   button_proofread_tooltip: "Proofread Writing"
   message_no_errors_found: "No writing errors were found.",
   message_server_error: "There was a problem communicating with the After the Deadline service. Try again in one minute."
};</pre>

Then make TinyMCE use these strings:

<pre>tinyMCE.addI18n('en.AtD', my_plugin_strings);</pre>

Note: I used _en.AtD_ to indicate _my_plugin_strings_ are an English translation of this AtD extension. For other languages use
_**lang**.AtD_ where _**lang**_ is the two letter language code value specified for _language_ in your TinyMCE init.

These string labels are compatible with the [AtD/jQuery extension](http://www.afterthedeadline.com/download.slp?platform=jQuery).

## License

[LGPL](http://www.opensource.org/licenses/lgpl-2.1.php)

This code is a hack on the spellcheck plugin from Moxiecode.  Thanks for the hard work guys.

## Contact

This code has always been open source.  We're putting it on Github so that you can feel free to fork it, hack it, and release your own version.

Join the [atd-developers](http://groups.google.com/group/atd-developers) list for community support.


CHANGELOG.md
### 24 Mar 10

*   Updated to latest AtD/Core module:

        *   Added L10n string _menu_title_confused_word_ to localize "Did you mean..."
    *   Fixed two cases of parent variable polution (two for loops not declaring their vars)
    *   Fixed bug preventing subsequent occurences of one error (w/ the same context) from highlighting
    *   Error highlighter now uses beginning of word boundary to accurately find error location in text
    *   Fixed bug preventing misspelled words in single quotes from being highlighted

### 15 Feb 10

*   Fixed a highlighting issue
*   Fixed I18n issue with AtD/Core module
*   Fixed a bug where AtD Core was ignoring show types setting

### 14 Jan 10

*   Made plugin localizable. Just need string translations now :)
*   Plugin now uses AtD Core UI module. Created a make.sh script to combine Core UI script and editor plugin src. Will also minify them if jsmin is in the current path.

### 11 Dec 09

*   Error highlighting is better at not highlighting a suggestion within a word
*   Plugin now checks if a global constant, AtD_proofread_click_count exists. If it does--it increments it with each use.

### 12 Nov 09

*   Updated docs/examples to reflect AtD support for diacritical marks (accented words)
*   Plugin is now smarter about determining if no writing errors were highlighted and notifies the user when appropriate.

### 3 Nov 09

*   Overhauled readme, change log, and added a demo.html file.
*   Fixed a bug where highlighted errors could step on eachother (sometimes).

### 8 Sept 09

*   More updates than you can shake a stick at.

### 18 Jun 09

*   (18 Jun 09) Added a atd_strip_on_get option.  Some applications like WordPress grab the editor
  content occasionally for an auto save.  In these cases the app plugin needs to
  strip the AtD tags and this option should be set to false.*   (19 Jun 09) fixed syntax error in IE and Safari (that FF chose to ignore)...

### 1 Jun 09

*   (31 May 09) updated plugin to treat -- as a separator.
*   (31 May 09) fixed plugin not turning off progress thingie when an AJAX error occurs
*   (29 May 09) fixed plugin not highlighting anything when string field of an error response is empty.
*   (25 May 09) fixed a bug causing underline and some other styles to get mangled in Wordpress.  Oops. :)
*   (24 May 09) fixed a bug with token text containing encoded HTML elements
  rather than the text its meant to represent (i.e. &quote; over a ").
  This caused highlighting to mess up in some cases.
*   (24 May 09) fixed a bug preventing mispelled words with a hyphen from getting  highlighted.
*   (24 May 09) fixed a bug where [hookword] [word 1] was causing other phrases with the
  pattern [hookword] [word 1][rest of word] to get highlighted
*   (17 May 09) multiword phrases now properly set the "previous" token used as a hook for
  checking if the next word/phrase has an error associated with it.  Naturally this fixes
  a bug where an error following a highlighted phrase isn't highlighted.

### 15 May 09

*   Rewrote the word highlighting code, the new stuff is easier to mantain and much cleaner. This plugin is no longer such a bad hack.  It's still a hack though.
*   Added the atd_ignore_strings option and the atd_ignore_enable option.

### Release 14 Apr 09

*   updated readme to show correct fom for atd_rpc_url.  The variables requires a ?url= at the end.



src/editor_plugin.js
/*
 * TinyMCE Writing Improvement Tool Plugin 
 * Author: Raphael Mudge
 * Discuss: https://groups.google.com/forum/#!forum/atd-developers
 *
 * http://open.afterthedeadline.com
 *
 * Distributed under the LGPL
 *
 * Derived from:
 *    $Id: editor_plugin_src.js 425 2007-11-21 15:17:39Z spocke $
 *
 *    @author Moxiecode
 *    @copyright Copyright (C) 2004-2008, Moxiecode Systems AB, All rights reserved.
 *
 *    Moxiecode Spell Checker plugin released under the LGPL with TinyMCE
 */

(function() 
{
   var JSONRequest = tinymce.util.JSONRequest, each = tinymce.each, DOM = tinymce.DOM; 

   tinymce.create('tinymce.plugins.AfterTheDeadlinePlugin', 
   {
      getInfo : function() 
      {
         return 
         ({
	    longname :  'After The Deadline',
            author :    'Raphael Mudge',
	    authorurl : 'http://blog.afterthedeadline.com',
	    infourl :   'http://open.afterthedeadline.com',
	    version :   tinymce.majorVersion + "." + tinymce.minorVersion
	 });
      },

      /* initializes the functions used by the AtD Core UI Module */
      initAtDCore : function(editor, plugin)
      {
         var core = new AtDCore();

         core.map = each;

         core.getAttrib = function(node, key) 
         { 
            return editor.dom.getAttrib(node, key); 
         };

	 core.findSpans = function(parent) 
         {
            if (parent == undefined)
               return editor.dom.select('span');
            else
               return editor.dom.select('span', parent);
         };

         core.hasClass = function(node, className) 
         { 
            return editor.dom.hasClass(node, className); 
         };
         
         core.contents = function(node) 
         { 
            return node.childNodes;  
         };

         core.replaceWith = function(old_node, new_node) 
         { 
            return editor.dom.replace(new_node, old_node); 
         };

         core.create = function(node_html) 
         { 
            return editor.dom.create('span', { 'class': 'mceItemHidden' }, node_html);
         };

         core.removeParent = function(node) 
         {
	    editor.dom.remove(node, 1);
            return node;
         };

         core.remove = function(node) 
         { 
            editor.dom.remove(node); 
         };

         core.getLang = function(key, defaultk) 
         { 
             return editor.getLang("AtD." + key, defaultk);
         };

         core.setIgnoreStrings(editor.getParam("atd_ignore_strings", ""));
         core.showTypes(editor.getParam("atd_show_types", ""));
         return core;
      },
 
      /* called when the plugin is initialized */
      init : function(ed, url) 
      {
         var t = this;
         var plugin  = this;
         var editor  = ed;
         var core = this.initAtDCore(editor, plugin);

         this.url    = url;
         this.editor = ed;
         ed.core = core;

         /* look at the atd_ignore variable and put that stuff into a hash */
         var ignore = tinymce.util.Cookie.getHash('atd_ignore');

         if (ignore == undefined)
         {
            ignore = {};
         }

         /* add a command to request a document check and process the results. */
         editor.addCommand('mceWritingImprovementTool', function(callback) 
         {
            /* checks if a global var for click stats exists and increments it if it does... */
            if (typeof AtD_proofread_click_count != "undefined")
               AtD_proofread_click_count++;

            /* create the nifty spinny thing that says "hizzo, I'm doing something fo realz" */
            plugin.editor.setProgressState(1);

            /* remove the previous errors */
            plugin._removeWords();

            /* send request to our service */
            plugin.sendRequest('checkDocument', ed.getContent({ format: 'raw' }), function(data, request, someObject)
            {
               /* turn off the spinning thingie */
               plugin.editor.setProgressState(0);

               /* if the server is not accepting requests, let the user know */
               if (request.status != 200 || request.responseText.substr(1, 4) == 'html')
               {
                  ed.windowManager.alert( plugin.editor.getLang('AtD.message_server_error', 'There was a problem communicating with the After the Deadline service. Try again in one minute.') );
                  return;
               }

               /* check to see if things are broken first and foremost */
               if (request.responseXML.getElementsByTagName('message').item(0) != null)
               {
                  ed.windowManager.alert(request.responseXML.getElementsByTagName('message').item(0).firstChild.data);
                  return;
               }

               var results = core.processXML(request.responseXML);
               var ecount  = 0;

               if (results.count > 0)
               {
                  ecount = plugin.markMyWords(results.errors);
                  ed.suggestions = results.suggestions; 
               }

               if (ecount == 0 && (!callback || callback == undefined))
                  ed.windowManager.alert(plugin.editor.getLang('AtD.message_no_errors_found', 'No writing errors were found.'));
               else if (callback)
                  callback(ecount);
            });
         });
          
         /* load cascading style sheet for this plugin */
     	 editor.onInit.add(function() 
         {
            /* loading the content.css file, why? I have no clue */
            if (editor.settings.content_css !== false)
            {
               editor.dom.loadCSS(editor.getParam("atd_css_url", url + '/css/content.css'));
            }
	 });

         /* again showing a menu, I have no clue what */
	 editor.onClick.add(plugin._showMenu, plugin);

         /* we're showing some sort of menu, no idea what */
	 editor.onContextMenu.add(plugin._showMenu, plugin);

         /* strip out the markup before the contents is serialized (and do it on a copy of the markup so we don't affect the user experience) */
         editor.onPreProcess.add(function(sender, object) 
         {
            var dom = sender.dom;

            each(dom.select('span', object.node).reverse(), function(n) 
            {
               if (n && (dom.hasClass(n, 'hiddenGrammarError') || dom.hasClass(n, 'hiddenSpellError') || dom.hasClass(n, 'hiddenSuggestion') || dom.hasClass(n, 'mceItemHidden') || (dom.getAttrib(n, 'class') == "" && dom.getAttrib(n, 'style') == "" && dom.getAttrib(n, 'id') == "" && !dom.hasClass(n, 'Apple-style-span') && dom.getAttrib(n, 'mce_name') == ""))) 
               {
                  dom.remove(n, 1);
               }
            });
         });

         /* cleanup the HTML before executing certain commands */
	 editor.onBeforeExecCommand.add(function(editor, command) 
         {
            if (command == 'mceCodeEditor')
            {
               plugin._removeWords();
            }
            else if (command == 'mceFullScreen')
            {
               plugin._done();
            }
         });
      },

      createControl : function(name, controlManager) 
      {
         var control = this;

         if (name == 'AtD') 
         {
            return controlManager.createButton(name, { 
               title: this.editor.getLang('AtD.button_proofread_tooltip', 'Proofread Writing'),
               image: this.editor.getParam('atd_button_url', this.url + '/atdbuttontr.gif'), 
               cmd: 'mceWritingImprovementTool', 
               scope: control 
            });
         }
      },

      _removeWords : function(w) 
      {
         var ed = this.editor, dom = ed.dom, se = ed.selection, b = se.getBookmark();

         ed.core.removeWords(undefined, w);

         /* force a rebuild of the DOM... even though the right elements are stripped, the DOM is still organized
            as if the span were there and this breaks my code */

         dom.setHTML(dom.getRoot(), dom.getRoot().innerHTML);

         se.moveToBookmark(b);
      },

      markMyWords : function(errors)
      {
         var ed  = this.editor;
         var se = ed.selection, b = se.getBookmark();

         var ecount = ed.core.markMyWords(ed.core.contents(this.editor.getBody()), errors);

         se.moveToBookmark(b);
         return ecount;
      },

      _showMenu : function(ed, e) 
      {
         var t = this, ed = t.editor, m = t._menu, p1, dom = ed.dom, vp = dom.getViewPort(ed.getWin());
         var plugin = this;

         if (!m) 
         {
            p1 = DOM.getPos(ed.getContentAreaContainer());
            //p2 = DOM.getPos(ed.getContainer());

            m = ed.controlManager.createDropMenu('spellcheckermenu', 
            {
               offset_x : p1.x,
               offset_y : p1.y,
               'class' : 'mceNoIcons'
            });

            t._menu = m;
         }

         if (ed.core.isMarkedNode(e.target))
         {
            /* remove these other lame-o elements */
            m.removeAll();

            /* find the correct suggestions object */
            var errorDescription = ed.core.findSuggestion(e.target);

            if (errorDescription == undefined)
            {
               m.add({title : plugin.editor.getLang('AtD.menu_title_no_suggestions', 'No suggestions'), 'class' : 'mceMenuItemTitle'}).setDisabled(1);
            }
            else if (errorDescription["suggestions"].length == 0)
            {
               m.add({title : errorDescription["description"], 'class' : 'mceMenuItemTitle'}).setDisabled(1);
            }
            else
            {
               m.add({ title : errorDescription["description"], 'class' : 'mceMenuItemTitle' }).setDisabled(1);

               for (var i = 0; i < errorDescription["suggestions"].length; i++)
               {
                  (function(sugg)
                   {
                      m.add({
                         title   : sugg, 
                         onclick : function() 
                         {
                            ed.core.applySuggestion(e.target, sugg);
                            t._checkDone();
                         }
                      });
                   })(errorDescription["suggestions"][i]);
               }

               m.addSeparator();
            }
             
            if (errorDescription != undefined && errorDescription["moreinfo"] != null)
            {
               (function(url)
                {
                   m.add({
                     title : plugin.editor.getLang('AtD.menu_option_explain', 'Explain...'),
                     onclick : function() 
                     {
                        ed.windowManager.open({
                           url : url,
                           width : 480,
                           height : 380,
                           inline : true
                        }, { theme_url : this.url });
	             }
                  });
               })(errorDescription["moreinfo"]);

               m.addSeparator();
            }

            m.add({
               title : plugin.editor.getLang('AtD.menu_option_ignore_once', 'Ignore suggestion'),
               onclick : function() 
               {
                  dom.remove(e.target, 1);
                  t._checkDone();
	       }
            });

            if (String(this.editor.getParam("atd_ignore_enable",  "false")) == "true")
            {
                m.add({
                  title : plugin.editor.getLang('AtD.menu_option_ignore_always', 'Ignore always'),
                  onclick : function() 
                  {
                      var url = t.editor.getParam('atd_ignore_rpc_url', '{backend}');

                      if (url == '{backend}')
                      {
                         /* Default scheme is to save ignore preferences in a cookie */

                         var ignore = tinymce.util.Cookie.getHash('atd_ignore'); 
                         if (ignore == undefined) { ignore = {}; }
                         ignore[e.target.innerHTML] = 1;
                  
                         tinymce.util.Cookie.setHash('atd_ignore', ignore, new Date( (new Date().getTime()) + 157680000000) );
                      }
                      else
                      {
                         /* Plugin is configured to send ignore preferences to server, do that */

                         var id  = t.editor.getParam("atd_rpc_id",  "12345678");

                         tinymce.util.XHR.send({
                             url          : url + encodeURI(e.target.innerHTML).replace(/&/g, '%26') + "&key=" + id,
                             content_type : 'text/xml',
                             async        : true,
                             type         : 'GET',
                             success      : function( type, req, o )
                             {
                                /* do nothing */
                             },
                             error        : function( type, req, o )
                             {
                                alert( "Ignore preference save failed\n" + type + "\n" + req.status + "\nAt: " + o.url ); 
                             }
                         });

                         /* update atd_ignore_strings with the new value */
                         t.editor.core.setIgnoreStrings(e.target.innerHTML); /* this does an update */
                     }

                     t._removeWords(e.target.innerHTML);
                     t._checkDone();
                  }
               });
            }
            else
            {
                m.add({
                  title : plugin.editor.getLang('menu_option_ignore_all', 'Ignore all'),
                  onclick : function() 
                  {
                     t._removeWords(e.target.innerHTML);
                     t._checkDone();
                  }
               });
            }

            /* show the menu please */
            ed.selection.select(e.target);
            p1 = dom.getPos(e.target);
            m.showMenu(p1.x, p1.y + e.target.offsetHeight - vp.y);

            return tinymce.dom.Event.cancel(e);
         } 
         else
         {
            m.hideMenu();
         }
      },

      /* loop through editor DOM, call _done if no mce tags exist. */
      _checkDone : function() 
      {
         var t = this, ed = t.editor, dom = ed.dom, o;

         each(dom.select('span'), function(n) 
         {
            if (n && dom.hasClass(n, 'mceItemHidden'))
            {
               o = true;
               return false;
            }
         });

         if (!o)
         {
            t._done();
         }
      },

      /* remove all tags, hide the menu, and fire a dom change event */
      _done : function() 
      {
         var plugin    = this;
         plugin._removeWords();

         if (plugin._menu)
         {
            plugin._menu.hideMenu();
         }

         plugin.editor.nodeChanged();
      },

      sendRequest : function(file, data, success)
      {
         var id  = this.editor.getParam("atd_rpc_id",  "12345678");
         var url = this.editor.getParam("atd_rpc_url", "{backend}");
         var plugin = this;

         if (url == '{backend}' || id == '12345678') 
         {
            this.editor.setProgressState(0);
            alert('Please specify: atd_rpc_url and atd_rpc_id');
            return;
         }
   
         tinymce.util.XHR.send({
            url          : url + "/" + file,
            content_type : 'text/xml',
            type         : "POST",
            data         : "data=" + encodeURI(data).replace(/&/g, '%26') + "&key=" + id, 
            async        : true,
            success      : success,
            error        : function( type, req, o )
            {
	       plugin.editor.setProgressState(0);
               alert( type + "\n" + req.status + "\nAt: " + o.url ); 
            }
         });
      }
   });

   // Register plugin
   tinymce.PluginManager.add('AtD', tinymce.plugins.AfterTheDeadlinePlugin);
})();
